### ISSUE TITLE
All components: Improve type support

### ISSUE THREAD

**Issue by septatrix** [primefaces/primevue#3995]
### Describe the feature you would like to see added

Vue 3.3 introduced generic components. Using these, it is now possible to provide precise, high-quality types for props, events and slots with generic parameters. (For props this was already possible though quite cumbersome.)

### Is your feature request related to a problem?

It is frustrating to deal with untyped or insufficiently typed components in an application. In such cases autocompletion stops working leading to worse DX but also, and for the resulting product more important, it is no longer possible to check correctness of the code.

This is most often the case when working with events, slots and complex components like data tables.

### Describe the solution you'd like

It would be great if this project could utilize the new generic capabilities from Vue 3.3. It is possible to use this within .d.ts files, however, in the long term it will likely be beneficial to consider moving to typescript which would automate a lot of this stuff and can also be used for automatic and precise API/documentation generation.

### Describe alternatives you have considered

Not adding generic support: This is the status quo which comes with drawbacks in DX and type safety.

### Additional context

Vue 3.3 announcement blog post: https://blog.vuejs.org/posts/vue-3-3#script-setup-typescript-dx-improvements

---

I am not sure if/how this can be applied to the current API for data tables where columns are their own component. Other components already have support for typed column definitions as they are passed to the table component and can thus be typed together with knowledge of the passed values. (This is also possible with earlier Vue versions though a bit more convoluted).
As the table and columns are different components in primevue I am not sure if this can be supported without API changes.

Regardless, it would still be possible to offer this support for other components.

**Comment by m3hari** [comment]
@tugcekucukoglu @cagataycivici 

What do you think about this request? It will really improve DX, we are struggling with slots type-safety in our project. 

I can do PR if you give a green light

**Comment by cwdx** [comment]
+1 for DataTable/DataView generics support

```ts
// DataTable.vue
<script setup lang="ts" generic="T">
defineProps<{
  value: T[]
}>()
</script>
```

```vue
<!-- MyComponent.vue -->
 <DataTable
    :value="rows"
/>
```

**Comment by louia** [comment]
Any news on this ? @tugcekucukoglu @cagataycivici

**Comment by hotip** [comment]
Any news?

**Comment by SakuraBurst** [comment]
Any news?

**Comment by some-user123** [comment]
I only can second/support this issue. It's such a pain, that all type information is lost when passing my data of type `T[]` into the data table as  `value`. `v-model:selection` is typed `any` but should rather be `T[]`. And when the same data is accessed in templates via `#body="{ data }">` , data is unfortunately `any` again 😔 Which is really a pain... I had many errors that typescript would have been able to catch early.

**Comment by TheFinn15** [comment]
Any news? Because its really pain when need to use data from slots and this components expect props `any[]` 💀

**Comment by jumika** [comment]
Any news?

**Comment by LeaderbotX400** [comment]
Honestly, this whole repository is full of a lot of things that seem to break types. I'm currently just trying to do the 'Timeline' (as its relatively simple) and I can't get anything to type properly.

**Comment by LeaderbotX400** [comment]
If anyone is interested, I've started to work on this [here](https://github.com/LeaderbotX400/primevue)

**Comment by tugcekucukoglu** [comment]
If you share your work with us, you will significantly contribute to open source. PrimeVue also needs contributions from the community. PRs are always welcome.

https://primevue.org/contribution/

**Comment by p-kuen** [comment]
I would also be happy to contribute improving types. Although it is not easy because the whole codebase uses Options API? Any plans to switch to composition api? It is the standard for Vue 3.

**Comment by sambruggeman** [comment]
I'd love to know if this is being looked at or not, it's sad that all type knowledge is lost, and creating a wrapper around the datatable/column components to try and get type safetly is not working for me.

**Comment by blouflashdb** [comment]
That there are still no typed slots after 2 years is sad.

**Comment by NoelDeMartin** [comment]
I've been looking into this, and something to keep in mind is that given the current API of the DataTable component, it's not possible to infer the type for the columns automatically. This happens because rows are not rendered with slots, but using child components instead.

If this is eventually fixed, we'll probably need to indicate the type of the items explicitly for every column:

```html
<DataTable :value="items">
	
	<!-- Here, the 'data' inside the body template is typed as any. -->
	<Column field="name">
		<template #body="{ data }">
			{{ data.name }}
		</template>
	</Column>

	<!-- We could pass a prop to set the types, such as "value-types"... -->
	<Column field="name" :value-types="items">
		<template #body="{ data }">
			{{ data.name }}
		</template>
	</Column>

	<!-- Or set the generic type explicitly. -->
	<!-- @vue-generic {ItemType} -->
	<Column field="name">
		<template #body="{ data }">
			{{ data.name }}
		</template>
	</Column>

</DataTable>
```

An alternative to this would be to refactor the way DataTables are rendered altogether, and use slots to render columns instead of child components. But that would be a pretty bit change, so I don't think it's likely to happen.

In short, I don't think fixing this problem is as easy as it seems. It involves making API design decisions, not just fixing some code.

**Comment by cwdx** [comment]
FYI for now I've been finding this the best middle ground:

```vue
<script setup lang="ts">
import { ref } from "vue";

type User = {
  id: string;
  name: string;
}

const items = ref<User[]>([/* ... The users */]);
</script>

<template>
<DataTable :value="items">
	<Column field="name">
		<template #body="{ data }: { data: User }">
			{{ data.name }}
		</template>
	</Column>
</DataTable>
</template>
```

**Comment by NoelDeMartin** [comment]
I've also kept iterating since I wrote my last comment, and I've found a solution that works without modifying the existing components :D. But it is a bit weird 🙈. Basically, this is what I do:

```vue
<template>
    <Table>
        <TableColumn field="name" v-slot="{ data }">
            {{ data.name }}
        </TableColumn>
    </Table>
</template>

<script setup lang="ts">
const items = [{ name: 'John Doe'}]; // This could be a ref, etc.
const { Table, TableColumn } = useDataTable(items); 
</script>
```

Doing that, the data is always fully typed without the need to repeat the type declaration in every column :).

The way this works internally is that in the end, the same PrimeVue components are rendered, but the typescript definition of that `useDataTable` functions makes this fully typed. I also had to implement custom Table and TableColumn components in order to use the `generic="T"` syntax, but that can probably be avoided with enough Typescript/PrimeVue knowledge. I was already using some custom components in order to encapsulate pagination and other features, so I didn't look much into doing it with pure Typescript. But it should be possible to just define it like this:

```ts
import { DataTable, Column } from 'primevue';

// Write your typescript magic here
type TableComponent<T> = '...';
type TableColumnComponent<T> = '...';

export function useDataTable<T>(items: T[] | Ref<T[]>) {
    return {
        Table: DataTable as TableComponent<T>,
        TableColumn: Column as TableColumnComponent<T>,
    };
}
